nncf/common/tensor_statistics/collectors.py

```diff
@@ -0,0 +1,226 @@
+
+ReductionShape = Tuple[int]
+
+
+class TensorStatisticCollectorBase(ABC):
+    """Collector estimate statistics at the quantization point based on the provided reduction shape."""
+
+    def __init__(self, reduction_shape: ReductionShape = None, num_samples: int = None):
+        self._reduction_shape = reduction_shape
+        self._enabled = True
+        self._collected_samples = 0
+        self._num_samples = num_samples
+
+    def register_input(self, x):
+        """Registers input tensor"""
+        if not self._enabled or \
+                self._num_samples is not None and self._collected_samples >= self._num_samples:
+            return x
+        if self._reduction_shape is None:
+            self._reduction_shape = {tuple(x.shape)}
+        self._register_input(x)
+        self._collected_samples += 1
+        return x
+
+    @abstractmethod
+    def _register_input(self, x):
+        pass
+
+    def get_statistics(self):
+        """Returns collected statistics if present"""
+        if self._collected_samples == 0:
+            raise StatisticsNotCollectedError()
+        return self._get_statistics()
+
+    @abstractmethod
+    def _get_statistics(self):
+        pass
+
+    def enable(self):
+        self._enabled = True
+
+    def disable(self):
+        self._enabled = False
+
+    def reset(self):
+        """Resets all the statistics in the collector"""
+        self._collected_samples = 0
+        self._reset()
+
+    @abstractmethod
+    def _reset(self):
+        pass
+
+    def collected_samples(self) -> int:
+        return self._collected_samples
+
+
+class StatisticsNotCollectedError(Exception):
+    """Raised when the statistics are not collected but requested."""
+
+
+class OnlineTensorStatisticCollector(TensorStatisticCollectorBase):
+    """Base class for collectors that collect statistics in online regime, without storing the data."""
+
+
+class OfflineTensorStatisticCollector(TensorStatisticCollectorBase):
+    """Collects statistics in offline regime by storing and aggregating data afterwards."""
+
+    def __init__(self, reduction_shape: ReductionShape = None, num_samples: int = None, window_size: int = None):
+        super().__init__(reduction_shape, num_samples)
+        self._samples = deque(maxlen=window_size)
+
+    def _reset(self):
+        self._samples.clear()
+
+
+class MinMaxStatisticCollector(OnlineTensorStatisticCollector):
+    """Collector estimates min of minimum values and max of maximum values."""
+
+    def __init__(self, use_abs_max: bool, reduction_shape: ReductionShape, num_samples: int = None):
+        super().__init__(reduction_shape, num_samples)
+        self._use_abs_max = use_abs_max
+        self._tensor_processor = None
+
+        self._min_values = None
+        self._max_values = None
+
+    def _register_input_common(self, x: NNCFTensor):
+        if not self._tensor_processor:
+            self._tensor_processor = x.tensor_processor
```

Comment:
I don't really like this idiom. I guess I missed it during one of the previous reviews, but should each `NNCFTensor` really carry a processor with it? This situation corresponds to strong coupling, which is undesirable, and to additional possibility of system ending up in an incorrect state. I'd rather that the correct tensor processor was instantiated using polymorphism during the collector creation, rather than dynamically taken from the current tensor to be registered; and also each `NNCFTensor` should not retain a ref to the corresponding processor but instead all of the tensors in a given run under a specific FW should use a single tensor processor object obtained by separate, non-tensor-specific means. @daniil-lyakhov what do you think?

Reply:
The problem is in MaskPropagation algo. Running pruning algorithm, Symbolic mask propagation performed first. It uses framework specific metatypes to match node to correspondent purning operation and use SymbolicTensorProcessor to propagate masks. Then second mask propagation is performed, in this time with framework specific tensor processor. Now, mask propagation algo has no information about tensor processor. To set this information one need to keep it as an attribute of mask propagation and then forward to each pruning operation, because pruning operations *has no information about tensor nature*. In case you want to replace (ugly) forwarding to polymorphism instantiation, thus you will need to specify two kinds of pruning operations: pruning operations with framework specific tensor and with symbolic tensor, which is huge overkill in my opinion

Reply:
The main idea behind such idiom is all common algorithm don't need any information about tensor nature. In case of statistic collectors, I think, it's more clear to specify tensor processor as attribute when initializing collector.

 But in case of pruning algo all logic of pruning should be refactored to use polymorphism

Reply:
Created ticket 71626 to address this out of the scope of the PR